### PR TITLE
feat(mls): send MLS message action FS-696

### PR DESCRIPTION
**Pull request by ghuMustafa**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-696" title="FS-696" target="_blank">Sub-taskFS-696</a>  [iOS] Create `SendMLSMessageAction`
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

- Created an action `SendMLSMessagesAction` for counting the number or remaining key packages for the self user.

- Created an action handler `SendMLSMessagesActionHandler` for `SendMLSMessagesAction`.

- Added the action handler to the `SendMLSMessagesActionHandlerTests`.

#### How to Test

- Added tests for SendMLSMessagesActionHandler
----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Sources/Request Strategies/MLS/Actions/SendMLSMessagesAction.swift [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            case .unknown(status: let status):
+                return "Unknown error (response status: \(status))"
+            }
+        }
+    }
+
+    // MARK: - Properties
+
+    public var resultHandler: ResultHandler?
+    public var body: String
```
Comment (johnxnguyen) [inline Sources/Request Strategies/MLS/Actions/SendMLSMessagesAction.swift:81]: Perhaps we should call this `mlsMessage` or similar?

File: Sources/Request Strategies/MLS/Actions/SendMLSMessagesAction.swift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        case missingLegalHoldConsent
+        case legalHoldNotEnabled
+        case mlsProposalNotFound
+        case mlsKeyPackageRefNotFound
+        case noConversation
+        case mlsStaleMessage
+        case mlsClientMismatch
+        case mlsUnsupportedProposal
+        case mlsUnsupportedMessage
+        case endpointNotAvailable
```
Comment (David-Henner) [inline Sources/Request Strategies/MLS/Actions/SendMLSMessagesAction.swift:40]:
```suggestion
        case endpointUnavailable
```

for consistency with the other actions

File: Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandler.swift [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        action: ActionHandler<SendMLSMessagesAction>.Action
+    ) {
+        var action = action
+
+        switch (response.httpStatus, response.payloadLabel()) {
+        case (201, _):
+            action.notifyResult(.success(()))
+
+        case (400, "mls-protocol-error"):
+            action.notifyResult(.failure(.mlsProtocolError))
```
Comment (David-Henner) [inline Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandler.swift:61]: you can use the [helpers](https://github.com/wireapp/wire-ios-data-model/blob/9e5f7274f0c4de645351b4ed92e1b065968d9e7b/Source/Model/EntityAction.swift#L127) such as
```
action.succeed()
action.fail(with: .mlsProtocolError)
```

File: Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandler.swift [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        default:
+            action.notifyResult(.failure(.unknown(status: response.httpStatus)))
+        }
+    }
+}
+
+extension SendMLSMessagesActionHandler {
+
+}
```
Comment (David-Henner) [inline Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandler.swift:101]:
```suggestion
```

File: Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandlerTests.swift [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    let body = "body"
+    var sut: SendMLSMessagesActionHandler!
+
+    override func setUpWithError() throws {
+        sut = SendMLSMessagesActionHandler(context: syncMOC)
+    }
+
+    override func tearDownWithError() throws {
+        sut = nil
+    }
```
Comment (David-Henner) [inline Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandlerTests.swift:33]: should we call `super` ?

File: Sources/Request Strategies/MLS/MLSRequestStrategy.swift [hunk 6]
```diff
@@ -31,7 +31,10 @@ public final class MLSRequestStrategy: AbstractRequestStrategy {
         withManagedObjectContext managedObjectContext: NSManagedObjectContext,
         applicationStatus: ApplicationStatus
     ) {
-        entitySync = EntityActionSync(actionHandlers: [])
+        entitySync = EntityActionSync(actionHandlers: [
+            SendMLSMessagesActionHandler(context: managedObjectContext)
+        ])
+
```
Comment (David-Henner) [inline Sources/Request Strategies/MLS/MLSRequestStrategy.swift:37]: SwiftLint warning :)

File: Sources/Request Strategies/MLS/Actions/SendMLSMessagesAction.swift (hunk 1 above)
Comment (ghuMustafa) [reply Sources/Request Strategies/MLS/Actions/SendMLSMessagesAction.swift]: Yes defiantly! Changed.

File: Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandler.swift (hunk 3 above)
Comment (ghuMustafa) [reply Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandler.swift]: Yes Its more simple and clean.

File: Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandler.swift (hunk 4 above)
Comment (ghuMustafa) [reply Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandler.swift]: removed.

File: Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandlerTests.swift (hunk 5 above)
Comment (ghuMustafa) [reply Sources/Request Strategies/MLS/Actions/SendMLSMessagesActionHandlerTests.swift]: Yes we should call super here.

File: Sources/Request Strategies/MLS/MLSRequestStrategy.swift (hunk 6 above)
Comment (ghuMustafa) [reply Sources/Request Strategies/MLS/MLSRequestStrategy.swift:37]: Removed

**Review by David-Henner — APPROVED** [review]

**Pull request merged**